blockprod/src/detail/block_maker.rs

```diff
@@ -207,46 +209,120 @@ impl BlockMaker {
+                self.current_tip_id,
+                tx_merkle_root,
+                witness_merkle_root,
+                timestamp,
+                consensus_data,
+            );
+
+            is_send(block_header.clone());
+
+            // TODO: find a way to use a oneshot channel. It doesn't seem to be supported in crossbeam.
+            let (result_sender, result_receiver) = crossbeam_channel::unbounded();
+            {
+                let chain_config = Arc::clone(&self.chain_config);
+                let current_tip_height = self.current_tip_height;
+                let stop_flag = Arc::clone(&stop_flag);
+                self.mining_thread_pool.spawn(move || {
+                    let block_header = Self::solve_block(
+                        chain_config,
+                        block_header,
+                        current_tip_height,
+                        stop_flag,
+                    );
+                    result_sender.send(block_header).unwrap();
+                });
             }
 
-            // attempt to receive new commands from the perpetual Block Builder
-            let new_info = match self.block_maker_rx.try_recv() {
-                Ok(cmd) => cmd,
-                Err(e) => match e {
-                    // if there's nothing from the channel, then we can keep trying to build the block
-                    crossbeam_channel::TryRecvError::Empty => continue,
-                    // if the channel is lost, that means the perpetual Block Builder is destroyed.
-                    // No point in continuing since it seems that the node exited.
-                    crossbeam_channel::TryRecvError::Disconnected => {
-                        log::error!("Block Maker control channel lost. Exiting Block Maker task on tip {} on best height {}", self.current_tip_id, self.current_tip_height);
-                        break;
-                    }
-                },
-            };
-
-            match new_info {
-                BlockMakerControlCommand::NewTip(block_id, _) => {
-                    // if there is a new tip, no point in continuing to mine this block
-                    if block_id != self.current_tip_id {
-                        break;
+            // In this we store the successfully mined block, if any.
+            // This is necessary because a no futures can be awaited inside `select!` macro
+            // TODO: is there a better way to get the value from the `select!` macro below instead of mutating this value?
+            let mut block_to_submit = None;
+
+            // We wait for one of the following to happen:
+            // 1. `self.block_maker_rx` to tell us that the chainstate has a new tip, or that we should stop for some other reason
+            // 2. the mining thread to finish, and tell us whether it was successful or not,
+            //    in which case we can attempt to submit the block to chainstate
+            crossbeam_channel::select! {
```

Comment:
Not sure we want to introduce more blocking code within an async function. I know that the current mining is blocking already, but I'm not sure if we gain anything here.